### ISSUE TITLE
spack prereq

### ISSUE THREAD

**Issue by shahzebsiddiqui** [LLNL/spack#25048]
It would be cool if spack can have a command like `spack check` to  run checks on a HPC system to verify [spack prerequisites](https://spack.readthedocs.io/en/latest/getting_started.html#prerequisites) are met. The `spack check` command is not sub-command so i think we can use this subcommand. 

```
siddiq90@cori04> spack check
==> Error: check is not a recognized Spack command or extension command; check with `spack commands`.
```

By default it will run the required checks at the facility to ensure pre-req are met in order to use spack properly. We want to make sure that commands like `wget`, `curl`, `tar`, `unzip`, `bzip2` are working. Some of these commands were not found on Spock system reported by @sameershende. 

### Rationale

A user can just git clone spack and run `spack check` on their system to ensure all the checks are passing. The tests should cover basic functionality of spack which we assume is provided by the Operating System however it may not be the case for instance if user is running spack in a container that is a minimal OS without development libraries you can get into weird issues. 

### Description

The output of `spack check` could be implemented using pytest if it makes sense with one-line per test that shows `PASS` or `FAIL` with one test per method. Let's say we have a test name **check_bzip2** which could run  `which bzip2`. The `spack check` command should exit with a non-zero returncode if any of the test fails. This can be useful if one was to add `spack check` as part of their script before running some commands. 

### General information

- [ ] I have run `spack --version` and reported the version of Spack
- [ ] I have searched the issues of this repo and believe this is not a duplicate

**Comment by adamjstewart** [comment]
@becker33 @alalazo @vsoch are currently working on allowing Spack to bootstrap many of these dependencies itself automatically, so this check may not be needed for much longer.

**Comment by vsoch** [comment]
I think that bootstrap PR is for style dependencies only.

it was suggested in slack that this could be done as an extension, @shahzebsiddiqui could you make an extension?

**Comment by adamjstewart** [comment]
@vsoch your PR is for style deps only, but we already bootstrap clingo automatically and I think @becker33 was working on other deps too.

**Comment by vsoch** [comment]
oh gotcha. Neat!

**Comment by alalazo** [comment]
See #22720 for clingo with buildcaches

**Comment by alalazo** [comment]
@shahzebsiddiqui Did you try prototyping this with [extension](https://spack.readthedocs.io/en/latest/extensions.html)? Also for the name of the command I think it would be better to use something more specific, like `spack prerequisites`. Finally, if it may be of help, you can find a table of prerequisites for Spack in this [repo](https://github.com/alalazo/spack-bootstrap-mirrors).

**Comment by shahzebsiddiqui** [comment]
So i have not used spack extensions this looks relatively new, though this may be worth a try. @sameershende uncovered some issues on Spock system like basic core utilities like `wget` were not found. So if i understand correctly, the extensions are run via `spack unit-test --extensions=<name>` which is using pytest under the hood to run the test but these may not be portable or applicable across sites. 

I am looking at the repo you shared @alalazo for spack prerequisites, couldn't we automatically run this as part of spack, i wouldn't mind changing the name though i have hard time typing `prerequisites` even though we have tab completion. Perhaps we could use `spack bootstrap` or `spack prereq` as the name, though i would think `spack bootstrap` may be misleading it may imply that spack is installing some things as part of its pre-req checks.  @tldahlgren can chime in we were discussing this in our ECP meeting, the idea was to have these tests builtin spack that is system agnostic that doesn't require facility to have tests for silly things like 

- `which wget`
- `which bzip2`
- `which tar` 

I would also think we would want to check for editor like (vi, vim, emacs, nano) which is typically installed in most Linux system just to make sure `spack edit <package>` works. It could be optional check to detect modules for instance running ``spack module [tcl|lmod]`` commands may trip out if one doesn't have a module system.

**Comment by vsoch** [comment]
@shahzebsiddiqui from the docs, it looks like the extensions can be run like commands, e.g.,

```bash
$ spack filter --help
```
The extension can also have tests, which is what you pointed out, but the extension is not explicitly run by the test.

**Comment by vsoch** [comment]
This seems pretty straight forward and fun, @shahzebsiddiqui I can put one together this weekend if you like.

**Comment by vsoch** [comment]
Tada!! The 10 minute sleepy project is alive!

https://github.com/vsoch/spack-prereq
https://asciinema.org/a/426954?speed=2

Something along these lines should work for you @shahzebsiddiqui ? Test it out, and update it and PR if you want any more functionality, or to add tests. Happy Friday!

**Comment by shahzebsiddiqui** [comment]
So is spack extension the only viable option for running arbitrary test at the facility. I was leaning on having this part of spack codebase or if spack has a repo for it's extension in it's main org. I think most of these checks should be pretty simple and if it's part of spack organization it can gain more visibility and perhaps we can get more contribution. 

I think we can run these sanity check as means as a quick system compatibility check. If it fails we will need to report the errors with the help desk. It would be good to have output of the errors and logfiles.

**Comment by vsoch** [comment]
This is probably outside of my decision making, but (my opinion) is that the checks for any specific center are going to be different enough that it wouldn't make sense to have in the production version of spack. But I could see creating an extension, and then if it's popular enough evaluating that again.

If you want explicit test cases with pytest, I didn't implement that (I just did a simple command that prints red/green statuses for each check and then exits on 0 given success or force pass) but you could implement it via tests too. I chose not to because that would require pytest.

**Comment by shahzebsiddiqui** [comment]
Great work @vsoch I saw the repo https://github.com/vsoch/spack-prereq and this looks great start . Yeah I think we should have some extension supported under the spack organization so that its easier to find examples right now it's fairly limited and this is a pretty new and powerful feature

**Comment by vsoch** [comment]
Sure I can transfer it there if that’s desired, let’s check with @alalazo and @tldahlgren first.

**Comment by adamjstewart** [comment]
I personally don't like Spack extensions and I agree with @shahzebsiddiqui. Spack core changes rapidly, and personal extensions quickly become out of date. In this case, checking for Spack's prereqs isn't site-specific and would benefit a lot of people. Of course, if Spack automatically bootstraps its deps, such a command won't be needed anymore.

**Comment by vsoch** [comment]
> Spack core changes rapidly, and personal extensions quickly become out of date.

I think the distinction here is allocation of responsibility - in spack core it becomes the responsibility of spack maintainers. An extension is not. The unspoken expectation is "I expect you to keep it working for me." I would argue this is reasonable for most of spack (or any open source project) for something that is a core feature, but people are not entitled to just request that anything and everything go in without having that proof first. This is why I think an extension is a reasonable approach to start - it will either prove that:

1. it's valued enough to be added, or 
2. prove that it's valued enough for a separate group of maintainers (perhaps @shahzebsiddiqui who is asking for it) to maintain separately.

> if Spack automatically bootstraps its deps, such a command won't be needed anymore.

And this would be the super ideal! But you probably need at least Python to be able to do that :P

**Comment by shahzebsiddiqui** [comment]
Right now the use-case was to decide whether E4S team should run basic sanity check for tests that depend on E4S Testsuite. I don't think these should be part of E4S Testsuite, bcz these are facility specific tests that can be addressed at the facility in their test suite. That is what buildtest,  reframe, or pavilion address. 

I think we should proceed with the extension for now what @vsoch has provided but we should move this to the spack organization and start to collect a few more extension so we get more traction. I think it's a good way to publicize spack extension, we should document the spack extension in the docs for ones officially supported by spack.  

Is it possible to have one repo to host all the spack extension or should we keep them separate. If I understand the spack extension need to have a specific directory structure.

**Comment by vsoch** [comment]
> I think we should proceed with the extension for now what @vsoch has provided but we should move this to the spack organization and start to collect a few more extension so we get more traction. 

I did this last night. 

> Is it possible to have one repo to host all the slack extension or is that not good idea.

I was thinking about this, and I have mixed feelings. On the one hand, it would be super nice to maintain a single extensions repository and then have common docs, place to post issues, etc., but on the other hand (with the way extensions are currently designed) that means that by adding the repository root you install them all. I think what would be ideal is to take an approach like [snakemake-wrappers](https://github.com/snakemake/snakemake-wrappers/) where they can all live in (and be tested, documented, etc.) in the same root, but then installed. E.g.,

```
$ spack install-extension prereq
``` 

Which would clone that repository, and then add _just_ the pre-req to some user or group centric folder of extensions. Then we can have our cake and it eat it too :)

**Comment by alalazo** [comment]
> Spack core changes rapidly, and personal extensions quickly become out of date.

@adamjstewart I don't think this is always the case. For instance, using `which` to detect executables is stable since a long time and here this is basically the functionality that gets used. Honestly I think extensions would be good in a number of ways to:
1. Deprecate features that have been superseded by new functionality (we can move them to an extension and remove them from core Spack). Think for instance about `spack activate/deactivate`.
2. Develop a new command without being blocked on core developers having to review your changes. With extensions you can be in charge of every decision and receive PRs/suggestions of your own without the risk of comments being lost in the 1k other issues that core Spack has. Also, nothing prevents commands that started as extensions to be incorporated in core Spack later on.

In general I see extensions as a scalable way to let people try new ideas based on core Spack and they don't sound to me too dissimilar to the "try your own scripts" which we promote during tutorials.

> Which would clone that repository, and then add just the pre-req to some user or group centric folder of extensions. Then we can have our cake and it eat it too :)

@vsoch Eheh, I tried something like that sometimes ago largely copying the `go` cli design, see #14181 It's currently stale since at the time it didn't receive a lot of traction. Maybe I should take advantage of the momentum here and revisit that PR :slightly_smiling_face:

**Comment by vsoch** [comment]
I would! I like the idea at least. :)

**Comment by shahzebsiddiqui** [comment]
I tested out this extension on my mac and it worked! Does it make sense to document the core spack extensions in https://spack.readthedocs.io/en/latest/extensions.html or do we document them in their separate projects. I noticed that @alalazo extension for `spack filter` was added in https://github.com/spack/spack-prereq/commit/bfc21bd4ecd9507dd54fe27ed8f447e651746207 but removed by @vsoch in https://github.com/spack/spack-prereq/commit/a0ce578d12befd096da8e3e85ed36d8300e57b5d. Are we going to keep these extensions separate or single project? 

I am happy with whatever you guys decide, I think my initial request for basic checks are met and we can close this ticket. Since this documented in https://github.com/spack/spack-prereq/blob/main/README.md i don't see point in documenting this spack documentation though it could be good to provide link to the spack extensions that are available:
 - https://github.com/spack/spack-prereq
 - https://github.com/alalazo/spack-scripting

**Comment by vsoch** [comment]
@shahzebsiddiqui I think you are getting confused just because the spack pre-req extension is based off of the filter extension and thus has a common commit history - I did this only to give @alalazo credit for most of that structure that he designed, and it doesn't indicate that repository is the "master extensions" repo that filter was removed from. There hasn't been discussion / decision about having some central spack extension repository afaik, so I wouldn't expect them to be in one place.

As you've already pointed out, if you want the fitler extension code it lives at: https://github.com/alalazo/spack-scripting.  I agree it would be useful to have a more central strategy for storing and then installing extensions, and I think probably the first thing to do is go back to @alalazo PR that attempts to make a common interface to install. Probably from there it would be logical to have a way to register known extensions to be available to that, regardless of the repository they live in. Once all that is designed, then yes it would make sense to update the docs for it.  But I would not choose a strategy that requires spack developers to manually add extensions - that is just more work. I'd do something akin to the snakemake workflow catalog: https://snakemake.github.io/snakemake-workflow-catalog/ (based on discovery) which I've reproduced for both [spack stacks](https://spack.github.io/spack-stack-catalog/) and [singularity containers](https://singularityhub.github.io/singularity-catalog/).

**Comment by shahzebsiddiqui** [comment]
I am going to close this now. I think this has been addressed and documented in the spack project to some extent at https://spack.readthedocs.io/en/latest/getting_started.html#system-prerequisites

**Comment by alalazo** [comment]
@shahzebsiddiqui We have a command for that too in `develop`:
```console
% spack bootstrap status
Spack v0.17.1 - python@3.8 

[PASS] Core Functionalities

[FAIL] Binary packages
  [B] MISSING "gpg2": required to sign/verify buildcaches


Spack will take care of bootstrapping any missing dependency marked as [B]. Dependencies marked as [-] are instead required to be found on the system.
```

**Comment by shahzebsiddiqui** [comment]
Thanks for sharing. I wasn't aware of this subcommand. I'll check out the docs.
